tensorflow_probability/python/mcmc/internal/util_test.py

```diff
@@ -196,25 +196,48 @@ def fn(x, y):
 @test_util.test_all_tf_execution_regimes
 class SmartForLoopTest(test_util.TestCase):
 
-  def test_python_for_loop(self):
+  def test_no_loop_iters(self):
     counter = None
     # Not @parameterized because the tf.constants would be executed outside the
     # Eager mode that @test_util.test_all_tf_execution_regimes creates, and
     # TF is unhappy about that.
-    for n in [10, tf.constant(10, dtype=tf.int64),
-              tf.constant(10, dtype=tf.int32)]:
+    for n in [0, tf.constant(0, dtype=tf.int64),
+              tf.constant(0, dtype=tf.int32)]:
       counter = collections.Counter()
       def body(x):
         counter['body_calls'] += 1
         return [x + 1]
 
       result = util.smart_for_loop(
           loop_num_iter=n, body_fn=body, initial_loop_vars=[tf.constant(1)])
-      expected_calls = 1 if JAX_MODE else 10  # JAX always traces loops
-      self.assertEqual(expected_calls, counter['body_calls'])
-      self.assertAllClose([11], self.evaluate(result))
+      if JAX_MODE:  # JAX always traces loops
+        self.assertEqual(1, counter['body_calls'])
+      else:
+        self.assertEqual(0, counter['body_calls'])
+      self.assertAllClose([1], self.evaluate(result))
+
+  @parameterized.parameters(1, 10)
+  def test_static_num_iters(self, iters):
+    counter = None
+    # following loop variables not @parameterized because the tf.constants
+    # would be executed outside the Eager mode that
+    # @test_util.test_all_tf_execution_regimes creates
+    for n in [iters, tf.constant(iters, dtype=tf.int64),
+              tf.constant(iters, dtype=tf.int32)]:
+      counter = collections.Counter()
+      def body(x):
+        counter['body_calls'] += 1
+        return [x + 1]
+
+      result = util.smart_for_loop(
+          loop_num_iter=n, body_fn=body, initial_loop_vars=[tf.constant(1)])
+      if tf.executing_eagerly() and not JAX_MODE:  # JAX always traces loops
+        self.assertEqual(iters, counter['body_calls'])
+      else:
+        self.assertEqual(1, counter['body_calls'])
```

Comment:
Would the following have the same test coverage and let you include the 0 case in `@parameterzed` instead of having to copy it separately?

```suggestion
      if JAX_MODE:  # JAX always traces loop bodies exactly once
        self.assertEqual(1, counter['body_calls'])
      elif tf.executing_eagerly():
        self.assertEqual(iters, counter['body_calls'])
      else:
        self.assertEqual(max(1, iters), counter['body_calls'])
```

Reply:
Yeah something slightly different works since 0 iterations means the function is never called even in graph mode (i.e. always `counter['body_calls'] = 0`). I wanted to keep them separate as the 0 case is special, but eliminating code duplication is also important.